server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/DropListRoutes.java

```diff
@@ -0,0 +1,217 @@
+    public static final String OWNER_SCOPE = ":ownerScope";
+    public static final String OWNER = ":owner";
+    public static final String DENIED_ENTITY = ":deniedEntity";
+    public static final String DENIED_ENTITY_TYPE = "deniedEntityType";
+
+    private final DropList dropList;
+    private final JsonTransformer jsonTransformer;
+
+    @Inject
+    public DropListRoutes(DropList dropList, JsonTransformer jsonTransformer) {
+        this.dropList = dropList;
+        this.jsonTransformer = jsonTransformer;
+    }
+
+    @Override
+    public String getBasePath() {
+        return DROP_LIST;
+    }
+
+    @Override
+    public void define(Service service) {
+        service.get(DROP_LIST + SEPARATOR + OWNER_SCOPE, this::getDropList, jsonTransformer);
+        service.get(DROP_LIST + SEPARATOR + OWNER_SCOPE + SEPARATOR + OWNER, this::getDropList, jsonTransformer);
+        service.put(DROP_LIST + SEPARATOR + OWNER_SCOPE + SEPARATOR + OWNER + SEPARATOR + DENIED_ENTITY, this::addDropListEntry);
+        service.put(DROP_LIST + SEPARATOR + OWNER_SCOPE + SEPARATOR + DENIED_ENTITY, this::addDropListEntry);
+        service.head(DROP_LIST + SEPARATOR + OWNER_SCOPE + SEPARATOR + OWNER + SEPARATOR + DENIED_ENTITY, this::dropListEntryExist);
+        service.head(DROP_LIST + SEPARATOR + OWNER_SCOPE + SEPARATOR + DENIED_ENTITY, this::dropListEntryExist);
+        service.delete(DROP_LIST + SEPARATOR + OWNER_SCOPE + SEPARATOR + OWNER + SEPARATOR + DENIED_ENTITY, this::removeDropListEntry);
+        service.delete(DROP_LIST + SEPARATOR + OWNER_SCOPE + SEPARATOR + DENIED_ENTITY, this::removeDropListEntry);
+    }
+
+    public ImmutableSet<String> getDropList(Request request, Response response) {
+        OwnerScope ownerScope = checkValidOwnerScope(request.params(OWNER_SCOPE));
+        String owner = Optional.ofNullable(request.params(OWNER)).orElse("");
+        Optional<DeniedEntityType> deniedEntityType = checkValidDeniedEntityType(request.queryParams(DENIED_ENTITY_TYPE));
+        Flux<DropListEntry> dropListEntries = dropList.list(ownerScope, owner);
+        if (deniedEntityType.isPresent()) {
+            dropListEntries = dropListEntries
+                .filter(deniedEntry -> deniedEntry.getDeniedEntityType().equals(deniedEntityType.get()));
+        }
+        return dropListEntries
+            .map(DropListEntry::getDeniedEntity)
+            .collect(ImmutableSet.toImmutableSet())
+            .block();
+    }
+
+    public String addDropListEntry(Request request, Response response) {
+        OwnerScope ownerScope = checkValidOwnerScope(request.params(OWNER_SCOPE));
+        String owner = Optional.ofNullable(request.params(OWNER)).orElse("");
+        String deniedEntity = request.params(DENIED_ENTITY);
+        DropListEntry dropListEntry = getDropListEntry(ownerScope, owner, deniedEntity);
+        dropList.add(dropListEntry)
+            .subscribeOn(Schedulers.boundedElastic())
+            .subscribe();
```

Comment:
.block

We want to be synchronous.

Reply:
If you don't mind, can you tell me why it is necessary to design a reactive API, but when calling it, use only blocking and be synchronous? After all, tasks to delete or add entities can work successfully through the scheduler

Reply:
Being reactive is:
 - overall a best practice
 - would shine when designing JMAP extensions for the user to manage his droplist as JMAP stack is reactive.
 - Future prone: if one day the matcher API becomes reactive or the webadmin becomes reactive, we will have less work.
 - Convention: all new storage we right is reactive.
 
 Of course APIs you worked with so far (webadmin, matcher) are not reactive so it is not that much of a good fit.
 
Now why being synchronous?
  - report failure to the user. 500 in case of failure. Not done with async...
  - flow control: synchronous flow control is easier to manage overall.
    
Webadmin framework is meant to be blocking so IMO let's block.

Is it clearer?

Reply:
Yes, now it's clear to me. Thank you very much for taking the time and answering my question in detail.